core/src/main/scala/org/apache/spark/storage/DiskStore.scala

```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
+
+  override def toByteBuffer(allocator: Int => ByteBuffer): ChunkedByteBuffer = {
+    val source = open()
+    try {
+      var remaining = blockSize
+      val chunks = new ListBuffer[ByteBuffer]()
+      while (remaining > 0) {
+        val chunkSize = math.min(remaining, Int.MaxValue)
+        val chunk = allocator(chunkSize.toInt)
+        remaining -= chunkSize
+
+        while (chunk.remaining() > 0) {
+          source.read(chunk)
+        }
+        chunk.flip()
+        chunks += chunk
+      }
+
+      new ChunkedByteBuffer(chunks.toArray)
+    } finally {
+      source.close()
+    }
+  }
+
+  override def size: Long = blockSize
+
+  override def dispose(): Unit = { }
+
+  private def open(): ReadableByteChannel = {
+    val channel = new FileInputStream(file).getChannel()
+    try {
+      CryptoStreamUtils.createReadableChannel(channel, conf, key)
+    } catch {
+      case e: Exception =>
+        Closeables.close(channel, true)
+        throw e
+    }
+  }
+
+  private class EncryptedManagedBuffer extends ManagedBuffer {
+
+    override def size(): Long = blockSize
+
+    override def nioByteBuffer(): ByteBuffer = {
+      // This is used by the block transfer service to replicate blocks. The upload code reads
+      // all bytes into memory to send the block to the remote executor, so it's ok to do this
+      // as long as the block fits in a Java array.
+      assert(blockSize <= Int.MaxValue, "Block is too large to be wrapped in a byte buffer.")
+      val is = toInputStream()
+      try {
+        ByteBuffer.wrap(ByteStreams.toByteArray(is))
+      } finally {
+        Closeables.close(is, true)
+      }
+    }
+
+    override def createInputStream(): InputStream = toInputStream()
+
+    override def convertToNetty(): Object = new ReadableChannelFileRegion(open(), blockSize)
+
+    override def retain(): ManagedBuffer = this
+
+    override def release(): ManagedBuffer = this
+
+  }
+
+}
+
+private class ReadableChannelFileRegion(source: ReadableByteChannel, blockSize: Long)
+  extends AbstractReferenceCounted with FileRegion {
+
+  private var _transferred = 0L
+
+  private val buffer = ByteBuffer.allocateDirect(64 * 1024)
+  buffer.flip()
+
+  override def count(): Long = blockSize
+
+  override def position(): Long = 0
+
+  override def transfered(): Long = _transferred
+
+  override def transferTo(target: WritableByteChannel, pos: Long): Long = {
+    assert(pos == transfered(), "Invalid position.")
+
+    var written = 0L
+    var lastWrite = -1L
+    while (lastWrite != 0) {
+      if (buffer.remaining() == 0) {
+        buffer.clear()
+        source.read(buffer)
+        buffer.flip()
+      }
+      if (buffer.remaining() > 0) {
+        lastWrite = target.write(buffer)
+        written += lastWrite
+      } else {
+        lastWrite = 0
+      }
+    }
+
+    _transferred += written
+    written
+  }
+
+  override def deallocate(): Unit = source.close()
```

Comment:
`StorageUtils.dispose` specifically checks for mapped buffers, which is not the case here. It could be changed, but in this case I wonder if it's necessary or if waiting for GC is good enough.